R/Quosure.R

```diff
@@ -0,0 +1,310 @@
+#' - if `object2` has unique code element placed before common element. This means that `object2`
+#' in the environment of the `object2` was evaluated some extra code before which can influence
+#' reproducibility
+#' - more cases todo
+#' @param object (`Quosure`)
+#' @param object2 (`Quosure`)
+#' @export
+setGeneric("join", function(object, object2) {
+  standardGeneric("join")
+})
+
+setMethod("join", signature("Quosure", "Quosure"), function(object, object2) {
+  # todo: revise chunks_merge_chunks and make robust join method
+
+  # object2 can't have modified object of the same name! See chunks_push_chunks
+  common_names <- intersect(ls(object@env), ls(object2@env))
+  is_overwritten <- vapply(common_names, function(x) {
+    !identical(get(x, object@env), get(x, object2@env))
+  }, logical(1))
+
+  # code
+  a_is_shared_code <- object@code %in% object2@code
+  b_is_shared_code <- object2@code %in% object@code
+
+  if (
+    any(a_is_shared_code) &&
+    (isFALSE(a_is_shared_code[1]) || isFALSE(b_is_shared_code[1]))
+  ) {
+    stop(
+      "`object` and `object2` can't be joined if one of the objects contains extra code before ",
+      "code shared by both."
+    )
+  } else if (
+    any(a_is_shared_code) && any(b_is_shared_code) &&
+    any(!a_is_shared_code) && any(!b_is_shared_code) &&
+    any(is_overwritten)) {
+    stop(
+      "`object` and `object2` have an extra code and environments contains modified object(s) of the same name ",
+      "Following object(s) have been modified:\n - ",
+      paste(common_names[is_overwritten], collapse = "\n - ")
+    )
+  }
+
+  # join expressions
+  # - remove duplicated code from object2 and append to the object1
+  unique_expr <- object2@code[!(object2@code %in% object@code & names(object2@code) %in% names(object@code))]
+  # - in case object@code and object2@code contains extra code with the same name - code2 needs to be renamed
+  object@code <- .keep_code_name_unique(object@code, unique_expr)
+
+  # insert (and overwrite) objects from env2 to env
+  object@env <- .copy_env(object@env)
+  lapply(ls(object2@env), function(x) assign(x, get(x, object2@env), object@env))
+
+  object
+})
+
+#' Copy environment content to the other environment
+#' To avoid sharing the same environment by multiplequo object (in reactives)
+#' @keywords internal
+.copy_env <- function(env) {
+  # need to copy the objects from old env to new env
+  # to avoid "chunks" sharing the same environment
+  env_new <- new.env()
+  for (i in ls(env)) {
+    env_new[[i]] <- env[[i]]
+  }
+  env_new
+}
+
+.keep_code_name_unique <- function(x, y) {
+  nm <- make.unique(c(names(x), names(y)))
+  setNames(c(x, y), nm)
+}
+
+
+reactive_quosure <- function(depends, expr) {
+  reactive(expr, env = depends()@env)
+}
+
+
+#' @export
+as_quosure <- function(x) {
+  UseMethod("as_quosure")
+}
+
+#' @export
+as_quosure.list <- function(x) {
+  new_quosure(list2env(x), code = as.expression(attr(x, "code")))
```

Comment:
as.expression for code is not needed as it will be automatically properly dispatch

Reply:
Yes, not needed, thanks. I put this once just to test something